controller/api/destination/watcher/pod_watcher.go

```diff
@@ -0,0 +1,543 @@
+// passed ip:port or its associated protocol config
+func (pw *PodWatcher) Unsubscribe(ip string, port Port, listener PodUpdateListener) {
+	pw.mu.Lock()
+	defer pw.mu.Unlock()
+
+	pw.log.Debugf("Stopping watch on %s:%d", ip, port)
+	pp, ok := pw.getPodPublisher(ip, port)
+	if !ok {
+		pw.log.Errorf("Cannot unsubscribe from unknown ip:port [%s:%d]", ip, port)
+		return
+	}
+	pp.unsubscribe(listener)
+
+	if len(pp.listeners) == 0 {
+		delete(pw.publishers, IPPort{pp.ip, pp.port})
+	}
+}
+
+// addPod is an event handler so it cannot block
+func (pw *PodWatcher) addPod(obj any) {
+	pod := obj.(*corev1.Pod)
+	pw.log.Tracef("Added pod %s.%s", pod.Name, pod.Namespace)
+	go pw.submitPodUpdate(pod, false)
+}
+
+// deletePod is an event handler so it cannot block
+func (pw *PodWatcher) deletePod(obj any) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			pw.log.Errorf("Couldn't get object from DeletedFinalStateUnknown %#v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*corev1.Pod)
+		if !ok {
+			pw.log.Errorf("DeletedFinalStateUnknown contained object that is not a Pod %#v", obj)
+			return
+		}
+	}
+	pw.log.Tracef("Deleted pod %s.%s", pod.Name, pod.Namespace)
+	go pw.submitPodUpdate(pod, true)
+}
+
+// updatePod is an event handler so it cannot block
+func (pw *PodWatcher) updatePod(oldObj any, newObj any) {
+	oldPod := oldObj.(*corev1.Pod)
+	newPod := newObj.(*corev1.Pod)
+	if oldPod.DeletionTimestamp == nil && newPod.DeletionTimestamp != nil {
+		// this is just a mark, wait for actual deletion event
+		return
+	}
+	pw.log.Tracef("Updated pod %s.%s", newPod.Name, newPod.Namespace)
+	go pw.submitPodUpdate(newPod, false)
+}
+
+func (pw *PodWatcher) submitPodUpdate(pod *corev1.Pod, remove bool) {
+	pw.mu.RLock()
+	defer pw.mu.RUnlock()
+
+	submitPod := pod
+	if remove {
+		submitPod = nil
+	}
+	for _, container := range pod.Spec.Containers {
+		for _, containerPort := range container.Ports {
+			if containerPort.ContainerPort != 0 {
+				if pp, ok := pw.getPodPublisher(pod.Status.PodIP, Port(containerPort.ContainerPort)); ok {
+					pp.updatePod(submitPod)
+				}
+			}
+			if containerPort.HostPort != 0 {
+				if pp, ok := pw.getPodPublisher(pod.Status.HostIP, Port(containerPort.HostPort)); ok {
+					pp.updatePod(submitPod)
+				}
+			}
+		}
+	}
+}
+
+// updateServer triggers an Update() call to the listeners of the podPublishers
+// whose pod matches the Server's selector. This function is an event handler
+// so it cannot block.
+func (pw *PodWatcher) updateServer(_ any) {
+	pw.mu.RLock()
+	defer pw.mu.RUnlock()
+
+	for _, pp := range pw.publishers {
+		if pp.pod == nil {
+			continue
+		}
+		opaquePorts := GetAnnotatedOpaquePorts(pp.pod, pw.defaultOpaquePorts)
+		_, isOpaque := opaquePorts[pp.port]
+		// if port is annotated to be always opaque we can disregard Server updates
+		if isOpaque {
+			continue
+		}
+
+		go func(pp *podPublisher) {
+			updated := false
```

Comment:
should we lock/unlock the mutex while in this go routine? Is it possible for a listener to get concurrently added/removed while we're going through the update logic in this routine?

Tried to put together an example here: https://go.dev/play/p/ADQ-EH3wVER

Reply:
Good catch, we're indeed missing a lock here :+1: